src/base/ClpMap.h

```diff
@@ -223,10 +223,13 @@ ClpMap<Key, Value, MemoryUsedBy>::erase(const IndexIterator &i)
 {
     assert(i != index_.end());
     const auto entryPosition = i->second;
+
     const auto sz = entryPosition->memCounted;
+    assert(memUsed_ >= sz);
+    memUsed_ -= sz;
```

Comment:
Moving this above the erase() violates the invariant that memUsed_ >= memory currently used for the duration of how long it takes to perform the erasures. Since those operations are far from atomic and may throw that introduces a memory leak opportunity.

Reply:
Updating the memory value before the erase() leads Squid to thinking it can add more entries when it should not. That is a memory leak behaviour.

I still believe it is best to erase in the order which minimizes potential side effects without having to rely on implementation details of the particular container used:
1) drop index entry. Leaves all data entry positions unaffected. Worst-case one entry has no index record.
2) drop data entry. Index is already updated successfully. Releases memory. Worst case memory accounting over-estimates usage.
3) update memory data. State becomes completely consistent again.

Yes, there are still bad cases with invariants. However with this ordering they all "fail safe" by preventing Squid consuming resources.

Reply:
> Updating the memory value before the erase() leads Squid to thinking it can add more entries when it should not.

No, it does not lead to that. Or, to be more precise, there is actually no "updating X before Y" here. The updates of all three interlinked components (memory consumption counter, search index, and storage) should be viewed as a single/atomic operation. From the invariants point of view, the order inside that operation does not matter. Moreover, no order can preserve correct state between component updates.

> I still believe it is best to erase in the order which minimizes potential side effects without having to rely on implementation details of the particular container used:

I do not think we can _meaningfully_ compare potential side effects of major invariant violations.

> 1. drop index entry

The "every stored entry can be found by its key" invariant is violated here.

> 2. drop data entry

The "every stored entry is accounted for in the memory consumption counter" invariant is violate here.

> 3. update memory data

Sanity restored.

> However with this ordering they all "fail safe" by preventing Squid consuming resources.

There is no true "fail safe" here. Each violation above may lead to innocent entries getting purged from the cache, consuming extra CPU cycles and other resources on obtaining fresh copies of that information.

Again, if you think it is better to reorder the steps, you should reorder them. I think the current order is the best because it preserves map state when an assertion is triggered.

Moreover, if you want to add an try/catch wrapper (around the code that should not throw unless we hit undefined behavior areas where all bets are off), I would not object.

Reply:
FTR, my understanding of "fail safe" definition for erase sequences like this is that during the "insanity" period other callers using the public API get(key) / find(key) / add(key) / del(key) should never be faced with risk of following an invalid iterator/pointer within the container. That means the index such callers interact with to find things needs to never point to missing/erased entry - even if some fluke occurrence causes the erase method to exit during "insanity".


Resolving this conversation, since we are now nitpicking edge cases that likely never occur.

Reply:
> my understanding of "fail safe" definition for erase sequences like this is that during the "insanity" period other callers using the public API

There must be no other callers during the critical section execution. It is virtually impossible to write correct code if there are such callers. If there is a possibility of such reentrant behavior that is not already covered by the existing assertions, then we should add more assertions.